pkg/cloud/cloud.go

```diff
@@ -674,40 +659,30 @@ func (c *cloud) WaitForAttachmentState(ctx context.Context, volumeID, expectedSt
 			return false, nil
 		}
 
-		if len(volume.Attachments) > 1 {
-			// Shouldn't happen; log so we know if it is
+		if volume.MultiAttachEnabled != nil && !*volume.MultiAttachEnabled && len(volume.Attachments) > 1 {
```

Comment:
Volume capabilities are checked in `ControllerPublishVolume` before `cloud.AttachDisk` is called. `MultiNodeMultiWriter` is only supported for block devices: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/9ea40839a594865efadf00d1a1455329b7bf0e79/pkg/driver/controller.go#L621

otherwise the `ControllerPublishVolume` request errors out.

Are you proposing enforcing checks beyond that? Open to those suggestions

Reply:
`isValidCapability` is called both in the attach path and detach path, specifically `ControllerPublishVolume` and `ControllerUnpublishVolume`. 

> I think it might make sense to check if the k8s-side volume has the MULTI_WRITER... capability as well, to prevent against accidents. I believe the capabilities should get passed into AttachVolume

I may be misunderstanding the suggestion, but we are already doing this in the controller service via `isValidCapability`.  If the request is to add access mode validation to the cloud service I'd recommend not doing so  to respect separation of concerns between these packages. We also only have access to `ctx`, `volumeId`, `nodeID` there.

Reply:
> `isValidCapability` is called both in the attach path and detach path, specifically `ControllerPublishVolume`

Nevermind then, I missed where that happens. If so, this can be ignored - resolving.